src/serde/de_tree.rs

```diff
@@ -0,0 +1,327 @@
+use std::convert::TryInto;
+use std::io::{Error, Read, Result, Write};
+
+use sha2::Digest;
+
+use crate::sha2::Sha256;
+
+use super::parse_atom::decode_size_with_offset;
+use super::utils::{copy_exactly, skip_bytes};
+
+const MAX_SINGLE_BYTE: u8 = 0x7f;
+const CONS_BOX_MARKER: u8 = 0xff;
+
+struct ShaWrapper(Sha256);
+
+impl Write for ShaWrapper {
+    fn write(&mut self, blob: &[u8]) -> std::result::Result<usize, Error> {
+        self.0.update(blob);
+        Ok(blob.len())
+    }
+    fn flush(&mut self) -> std::result::Result<(), Error> {
+        Ok(())
+    }
+}
+
+/// This data structure is used with `parse_triples`, which returns a triple of
+/// integer values for each clvm object in a tree.
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParsedTriple {
```

Comment:
This is a port of the python code. The name is pretty lousy, yeah. It's to emphasize the fact that what comes out to python is a triple of three integers, one of which has a split meaning based on whether `buffer[start]` is `0xff` or not (and that's how we get around storing whether it's an atom or pair).

Maybe `SubtreeOffset` or `ClvmSubtreeOffset` or `ParsedOffset` or something like that? Or `Position` or `Pos` instead of `Offset`, like what `seek` calls it?

Reply:
> one of which has a split meaning based on whether buffer[start] is 0xff or not (and that's how we get around storing whether it's an atom or pair).

That would be great to explain in a comment. But it also means you can get away without this distinction on the rust side (and save even more space)

> Maybe SubtreeOffset or ClvmSubtreeOffset or ParsedOffset or something like that?

It's not just an offset, it's a (non-owning) reference to a full CLVM object.

Reply:
Maybe something like `DeserializationOffset`